samples/hipRTC_gemm.cpp

```diff
@@ -0,0 +1,414 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <iostream>
+#include <vector>
+
+#include <hip/hip_runtime.h>
+#include <hip/hiprtc.h>
+
+#include <rocwmma/rocwmma.hpp>
+
+#include "common.hpp"
+// #include "hipRTC_gemm_kernel.hpp"
+
+using rocwmma::float16_t;
+using rocwmma::float32_t;
+using rocwmma::float64_t;
+
+#define HIPRTC_CALL(call) assert(call == HIPRTC_SUCCESS);
+#define HIP_CALL(call) assert(call == hipSuccess);
+
+// Host GEMM validation
+__host__ void gemm_cpu_h(uint32_t         m,
+                         uint32_t         n,
+                         uint32_t         k,
+                         float16_t const* a,
+                         float16_t const* b,
+                         float32_t const* c,
+                         float32_t*       d,
+                         uint32_t         lda,
+                         uint32_t         ldb,
+                         uint32_t         ldc,
+                         uint32_t         ldd,
+                         float32_t        alpha,
+                         float32_t        beta)
+{
+    for(int i = 0; i < m; ++i)
+    {
+        for(int j = 0; j < n; ++j)
+        {
+            float32_t accum = 0.0f;
+            for(int h = 0; h < k; ++h)
+            {
+                accum += static_cast<float32_t>(a[i * lda + h])
+                         * static_cast<float32_t>(b[j * ldb + h]);
+            }
+            d[i * ldd + j] = alpha * accum + beta * c[i * ldc + j];
+        }
+    }
+}
+
+// Matrix data initialization
+template <typename DataT>
+__host__ static inline void fill(DataT* mat, uint32_t m, uint32_t n)
+{
+    auto ld = n;
+    for(int i = 0; i < m; ++i)
+    {
+        for(int j = 0; j < n; ++j)
+        {
+            // Ascending order for each neighboring element.
+            // Alternate sign for even / odd
+            auto value      = (i * n + j) % 13;
+            mat[i * ld + j] = (value % 3) ? -static_cast<DataT>(value) : static_cast<DataT>(value);
+        }
+    }
+}
+
+// Supports ROCWMMA_M/N square sizes of
+// : 16 x 16
+// : 32 x 32
+const int ROCWMMA_M = 16;
+const int ROCWMMA_N = 16;
+
+// Supports ROCWMMA_K sizes as
+// : multiples of 16.
+const int ROCWMMA_K = 16;
+
+// AMDGCN default wave size
+const int WAVE_SIZE = rocwmma::AMDGCN_WAVE_SIZE;
```

Comment:
This needs to come from the device runtime. Try using the functionality in common.hpp